File: Problems/0200_0299/0284_Peeking_Iterator/Project_Java/PeekingIterator.java
import java.util.*;

// Java Iterator interface reference:
// https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html

class PeekingIterator implements Iterator<Integer> {
    // 4ms
    private Integer cache = null;
    private Iterator<Integer> myIterator;

    public PeekingIterator(Iterator<Integer> iterator) {
        // initialize any member here.
        myIterator = iterator;
        if (myIterator.hasNext())
            cache = myIterator.next();
    }

    // Returns the next element in the iteration without advancing the iterator. 
    public Integer peek() {
        return cache; 
    }

    // hasNext() and next() should behave the same as in the Iterator interface.
    // Override them if needed.
    @Override
    public Integer next() {
        Integer res = cache;
        cache = myIterator.hasNext() ? myIterator.next() : null;
        return res; 
    }

    @Override
    public boolean hasNext() {
        return cache != null;
    }
}

File: Problems/0200_0299/0284_Peeking_Iterator/Project_Go/solution/PeekingIterator.go
package solution

// 0ms

type PeekingIterator struct {
	myIterator []int
	cache int
}

func Constructor(iter *Iterator) *PeekingIterator {
	var this PeekingIterator
	for iter.hasNext() {
		this.myIterator = append(this.myIterator, iter.next())
	}	
	this.cache = 0
	return &this
}

func (this *PeekingIterator) hasNext() bool {
	return this.cache < len(this.myIterator)
}

func (this *PeekingIterator) next() int {
	this.cache++
	return this.myIterator[this.cache - 1]
}

func (this *PeekingIterator) peek() int {
	return this.myIterator[this.cache]
}
